### DISCUSSION TITLE
Upgrade from Qt4 to Qt6 for OpenFEDEM

### DISCUSSION THREAD

**Discussion by oeysstra** [openfedem/fedem-gui#8] [General]
As we strive to keep OpenFEDEM up-to-date with the latest technologies and maintain its robustness, we are planning to upgrade our graphical user interface (GUI) framework from Qt4 to Qt6. This upgrade will bring numerous benefits, including enhanced performance, improved features, and better support for modern development environments.

Key advantages of upgrading to Qt6 include:

- **Enhanced Performance:** Improved rendering and hardware acceleration capabilities.
- **Modern Features:** Access to new widgets and modules not available in Qt4.
- **Long-Term Support:** Continued updates and community support, ensuring future compatibility and security.

We understand that this transition may involve significant changes to the codebase and development workflows. Therefore, we are seeking input from the community to help plan and execute this upgrade smoothly. Your insights, experiences, and contributions will be invaluable in ensuring that the upgrade to Qt6 enhances OpenFEDEM without disrupting its usability or stability.

**Comment by kmokstad** [comment]
The branch [kmokstad/qt6-port](https://github.com/kmokstad/fedem-gui/tree/qt6-port) on my fork contains ongoing work for porting the GUI code to Qt6. This builds and runs successfully on Ubuntu 22.04 (with gcc 11), following the build instructions given [here](https://github.com/kmokstad/fedem-gui/blob/qt6-port/BUILD.md#build-fedem-gui-application-on-linux). Some remaining issues to be fixed:
- The curve plotting widget (based on Qwt) has not yet been ported.
- Creating curve by drag-and-drop does not work.
- The object tree in the left side of the main window lacks the lines indicating sub-object relationships.
- Many of the widgets have improper size such that they might overlap (this is not related to Qt version though).

**Reply by kmokstad** [reply]
The first item above is fixed, by linking Qwt 6.3 as an external library. The BUILD.md file is updated with the necessary details. However, another non-working item:

- The solver processes do not start - maybe a QProcess porting issue.

**Comment by kmokstad** [comment]
The [qt6-port](https://github.com/openfedem/fedem-gui/tree/qt6-port) branch has now been rebased on top of the main branch, such that it is up to date with R8.0.8. The issues above are still unresolved though, so still some work before this can be merged.

**Comment by kmokstad** [comment]
The above mentions issues (and more) have been resolved, and the first release based on Qt 6 is now here (R8.1.0).
Thereby closing this discussion as resolved.
